python/cudf/cudf/core/dataframe.py

```diff
@@ -164,6 +164,8 @@ def __init__(self, data=None, index=None, columns=None):
 
         self._add_empty_columns(columns, index)
 
+        self._enable_setattr_setitem = True
```

Comment:
Is there ever a situation we want this set to False?

Reply:
In the initializer prior to adding, for example, private attributes. Otherwise you couldn’t set them.

Reply:
The alternative is “registering” all attributes explicitly, and checking that list inside the dunder.

Reply:
This is a runtime flag - basically it says that `__init__` has completed, and future `__setattr__` attempts will have to pass through the new logic.

Should it be named `_init_complete` or similar to reflect that?

Reply:
I'm fairly opposed to the whitelist solution, so I like this.

Reply:
I was thinking about how we could create new attributes after-the-fact even when the names correspond to an existing column, and thought that the context management protocol would work well. Something like...

```
df["a"] = [0, 1, 2]
with df.attr_lock:
  df.a = "attribute"

print(df.a) # prints "attribute"
```
where `attr_lock` is a property which returns an object supporting the context management protocol. Under the hood it was just disable and re-enable the `_enable_setattr_setitem` flag.

Just a thought. Not looking to add something like that right now.

Reply:
I thought about naming the flag `_init_complete`, but decided against it because it was less specific, and assumed the flag would never be unset.